<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>IgQUAST by IgReC team</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name" align=center>IgQUAST</h1>
      <h2 class="project-tagline" align=center>A tool for full-length antibody/TCR repertoires quality assessment</h2>
      IgQUAST is a part of <a href="index.html" style="color:white;font-weight:bold;text-decoration: underline">IgReC</a> toolkit.
      Use the <a href="index.html" style="color:white;font-weight:bold;text-decoration: underline">main page</a> for download
    </section>

    <section class="main-content">
        <h2 id = "motivation">Motivation and purposes</h2>
        <p align = "justify">
            Construction of antibody repertoire is a preliminary step of any immunological analysis based on
            Rep-seq reads.
            During several last years at least three tools
            (IgRepertoireConstructor by Safonova et al (2015), MiXCR by Bolotin et al. (2015) and pRESTO by Vander Heiden et al. (2014))
            were released for this problem.
            However, for this moment there is certain leader among them.
            Each repertoire construction approach has its own advantages and disadvantages.
            In order to benchmark different approaches, we developed IgQUAST (<b>I</b>mmuno<b>G</b>lobulin <b>QU</b>ality
            <b>AS</b>sesstment <b>T</b>ool), a tool for benchmarking of full-length adaptive immune
            repertoire construction tools and quality assessment of adaptive immune repertoires.
        </p>

        <p align = "justify">
        <!-- <h3>IgQUAST purposes</h3> -->
        IgQUAST has the following purposes:
            <ol>
                <li> Reference-based and reference-free repertoire quality assessment;</li>
                <li> Overcorrection/undercorrection detection;</li>
                <li> Analysis of antibody abundances;</li>
                <li> Detection of repertoire construcion errors positions.</li>
            </ol>
        </p>


        <h2 id = "inputoutput">Quality assessment scenarios</h2>
        TODO add definition of repertoire.

        <p align = "justify">
        See detailed input requirements in <a href="igquast_manual.html#igquast_input">IgQUAST manual.</a>
        Please note, IgQUAST can benchmark repertoire construction stage itself only.
        It supposes that input reads are already merged, filtered from contaminations and cropped by V/J ends.
        </p>

        <p align = "justify">
        As an output IgQUAST reports a bunch of plots
        along with numerical statistics.
        All stats are reported in machine-readable JSON format in order to simplify automatic analysis.
        </p>

        <p align = "justify">
        For comprehensive analysis IgQIUAST also requires additional information: the initial read library and
        reference repertoire.
        IgQUAST can perform analysis using different scenarios depending on the provided input (described in the table below).
        </p>


        <p align = "justify">
        <table>
        <tr>
            <td align="center">
                <b>Scenario</b>
            </td>
            <td align="center">
                <b>Requirements</b>
            </td>
            <td align="center">
                <b>Resultant plots and statistics</b>
            </td>
        </tr>
        <tr>
            <td>
                <a href="#repertoire_to_repertoire_matching">Repertoire-to-repertoire matching</a>
            </td>
            <td>
                Constructed and reference repertoires in
                <a href="ig_repertoire_constructor_manual.html#clusters_fasta">CLUSTER.FA format</a>
            </td>
            <td>
            Constructed-to-reference and reference-to-constructed error distribution,
            error-position detection,
            multiplicity distribution,
            sensitivity-precision (optimal min-size estimation),
            cluster abundances estimation
            </td>
        </tr>
        <tr>
            <td>
                <a href="#reference_based_over_under_corection">Reference-based over/under correction detection</a>
            </td>
            <td>
                Initial input Rep-Seq library, reference and constructed repertoires
                <a href="ig_repertoire_constructor_manual.html#read_cluster_map">RCM</a> files
            </td>
            <td>
                Reference and constructed clusters purity and discordance distribution
            </td>
        </tr>
        <tr>
            <td>
                <a href="#reference_free_over_under_corection">Reference-free over/under correction detection</a>
                for constructed repertoire
            </td>
            <td>
                Initial input Rep-Seq library, constructed repertoire
                <a href="ig_repertoire_constructor_manual.html#read_cluster_map">RCM</a> file
            </td>
            <td>
                Error and discordance profile (for entire repertoire and by clusters),
                error distribution,
                suspicious cluster detection
            </td>
        </tr>
        <tr>
            <td>
                Over/under correction detection for reference repertoire</a>
            </td>
            <td>
                Initial input Rep-Seq library, reference repertoire
                <a href="ig_repertoire_constructor_manual.html#read_cluster_map">RCM</a> file
            </td>
            <td>
                The same results as for constructed repertoire
            </td>
        </tr>
        </table>
        </p>

        <p align = "justify">
            Some scenarios may be disabled by default since they are time- and memory-consuming.
            See <a href="igquast_manual.html#usage">IgQUAST manual</a> for options description.
        </p>

        <p align = "justify">
        If some required input is missing, IgQUAST tries to reconstruct missing info.
        For example, if some analysis requires RCM file for the repertoire, IgQUAST performs
        RCM restoration by mapping input reads into provided repertoire sequences.
        However, such restorations are not fully reliable since they require additional assumptions
        on the benchmarked tool. Thus, it is better to provide to IgQUAST as much information as possible.
        </p>

        <h2 id = "reference">Reference repertoire</h2>
        <p align = "justify">
        To obtain reference repertoire we
        have two options: simulate the repertoire or construct the repertoire using
        barcoded data and run repertoire construction in
        the <i>blind</i> (ignoring barcodes) mode.
        </p>

        <h3 id = "simulation">Simulation</h3>
        <p align = "justify">
        There are no special requirements for simulation software or/and simulated repertoire,
        but we suggest to use our <a href = 'http://yana-safonova.github.io/ig_simulator/'>IgSimulator</a>,
        that was specially designed for benchmarking adaptive immune repertoire construction strategies.
        </p>

        <h3 id = "umi">Molecular barcoding</h3>
        <p align = "justify">
        Molecular identifiers (UMIs) or <i>barcodes</i> can be used for B-cell labeling.
        Such labeling allows one to reconstruct adaptive immune sequences using consensus by groups
        and, thus,
        provides one with a reference dataset for benchmarking.
        </p>

        <p align = "justify">
        Please note that currently available barcoding techniques are not perfect.
        Thus, using naive barcode demultiplexing approach may lead to imprecise reference and biased analysis.
        The better option is to use an approach specially designed for barcode-assisted antibody repertoire construction.
        We strongly suggest to use our tool <a href="barigrec.html">BarIgReC</a> for this task.
        </p>

        <p align = "justify">
        Another approach to deal with barcoding artifacts is to consider specially designed semi-artificial test datasets.
        Such datasets could be obtained by aggressive input read and barcode filtering, when
        every suspicious barcode and every suspicious read is omitted.
        Such approach is definitely
        improper for real-life repertoire construction (since it loses a significant part of repertoire diversity, i.e., it is biased),
        but it is good for test dataset preparation.
        </p>


        <h2 id = "repertoire_to_repertoire_matching">Repertoire-to-repertoire matching</h2>
        For each reference repertoire sequence IgQUAST finds the closest sequence in the constructed repertoire
        and <i>vice versa</i>.
        This allows us to compare repertoires in terms of their sequences and detect typical positions of errors.

        ToDo Add story about reference threshold.

        <h3 id = "matching">Errors distribution</h3>
        For each sequence in the constructed repertoire IgQUAST finds the closest sequence in the reference repertoire
        and plots histogram of the distances. Also IgQUAST plots the same histogram mapping the reference repertoire to
        the constructed one. Thus, we obtain two kinds of plots: constructed-to-reference and reference-to-constructed
        errors distribution.

        Plot here.
        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/fuzzy2.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>

        We plot these plots for different <i>constructed min-size</i> values (1, 3, 5, 10).
        Altogether they form octoplot:

        Octoplot here!!!!!
        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/fuzzy1.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>

        These plots show how close input repertoires are.

        It is necessary to discover plots for different <i>min-size</i> since small clusters are probably imperfect and
        should be excluded from the analysis.

        <h3 id = "matching">Detection of repertoire construction errors positions</h3>

        For "almost correct" (matched with only one error) constructed clusters IgQUAST plots error-position distribution (error profile).

        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/fuzzy2.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>

        For example above analysis revealed that
        error are concentrated in FR4.
        The key reason of this is high error-rate at the end of Ig read.
        Fortunately, it could be easily fixed using J gene germline (e.g. <a>IMGT</a>) database.
        FR4 is a conservative region and we do not miss many actual variations.

        Our alignment tool VJFinder automatically fixes FR1 and FR4 during Ig reads alignment.
        Thus, this property is not an issue for IgReC.


        <h3 id = "matching">Sensitivity and precision</h3>
        IgQUAST computes clusters in the reference repertoire that were ideally recovered in the constructed repertoire.
        As integral measures of quality IgQUAST reports two following metrics:
        <ul>
            <li> <b>Precision</b> The larger precision, the less number of false constructed clusters</li>
            <li> <b>Sensitivity</b> The larger sensitivity, the more correct clusters were reconstructed</li>
        </ul>
        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/sensitivity_precision_definition.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>

        These metrics are <i>antagonistic:</i> generally, more sensitivity means less precision and vice versa.
        Thus, it is important to control both of them simultaneously.

        
        The story about min-size.
        Also add story about difference between two thresholds.
        <ul>
            <li> In case of real analysis we should exclude small clusters from the
                constructed repertoire to improve precision. Constructed min-size
                threshold is a parameter of the repertoire construction procedure</li>
            <li> For benchmarking we should not take small clusters into account during
                sensitivity/precision computation to make the metrics stable. Reference
                min-size threshold is a parameter of the benchmarking procedure</li>
            <li>IgQUAST computes sensitivity and precision for various constructed min-size
                values and for fixed reference min-size = 5</li>
        </ul>


        <h3 id = "matching">Finding optimal min-size threshold</h3>
        Distribution of cluster sizes is highly uneven!
        Thus, most clusters in a repertoire have small size.
        At the same time, consensus computed by a small cluster is unreliable


        We cannot construct consensus by two reads in non-trivial case
        Due to dependency of amplification errors larger cluster may be unreliable too


        <table align = center>
            <tr>
                <td align = center>
                    Add figure with different abundances and threshold effect demonstration.
                </td>
                <td align = center>
                    <img src = "docs/igquast/sens_prec_plot.png" width = 50% alt = "Sensitivity-precision plot example">
                </td>
            </tr>
        </table>

        

        <h3 id = "matching">Cluster abundances estimation</h3>
        Clusters are also characterized by their abundances.
        For clusters presented in both repertoires, IgQUAST compares abundances.
        Error correction does not fix ALL errors, hence typically cluster sizes are underestimated.
        I.e., we can expect that the cluster abundance in the constructed repertoire is less than abundance
        in the reference repertoire.
        Otherwise, if the constructed abundance is much more than the reference abundance
        (we say, "the abundance is overestimated"),
        we suppose that the correspondent cluster is overcorrected, i.e. erroneously glued with the other (probably smaller)
        cluster.

        In the plot possibly overcorrected clusters are shown by red points, while other cluster are shown by blue ones.


        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/underestimation1.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
                <td align = center>
                    <img src = "docs/igquast/underestimation2.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>


        <h2 id = "reference_based_over_under_corection">Reference-based over/undercorrection detection</h2>
        <h3 id = "matching">Cluster purity metrics</h3>
        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/purity.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>


        Constructed purity: measure of cluster overcorrection
        Reference purity: measure of cluster undercorrection


        Add plots. Mb. from slides.


        <h2 id = "reference_free_over_under_corection">Reference-free over/undercorrection detection</h2>
        <h3>Reference-free detection of over/undercorrected clusters</h3>

        To detect overcorrection construct consensus and count errors by position
        

        Remove story about splitting.


        <table align = center>
            <tr>
                <td align = center>
                    <img src = "docs/igquast/splitting1.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
                <td align = center>
                    <img src = "docs/igquast/splitting2.png" width = 50% alt = "Sensitivity and precision definition">
                </td>
            </tr>
        </table>

        <h2 id="feedback">Feedback and bug reports</h2>
        Your comments, bug reports, and suggestions are very welcome. They will help us to further improve IgQUAST.
        <br><br> If you have any trouble running IgQUAST, please provide us the log file from the output directory.
        <br><br> Address for communications: <a href="mailto:igtools_support@googlegroups.com">igtools_support@googlegroups.com</a>.


        <h2 id = "manuals">Manuals</h2>
        <p align = "center">
            <ul>
                <li>IgQUAST <a href = "igquast_manual.html">manual</a></li>
                <li>IgReC <a href = "ig_repertoire_constructor_manual.html">manual</a></li>
                <li>barcodedIgReC <a href = "barcodedIgReC.html">manual</a></li>
            </ul>
        </p>

        <h2 id = "citations">Citation</h2>
        <p align = "justify">
            <!-- <b>Citation:</b> -->
            Alexander Shlemov, Sergey Bankevich, Andrey Bzikadze,
            Dmitriy Chudakov,
            Yana Safonova,
            and Pavel A. Pevzner.
            Reconstructing antibody repertoires from error-prone immunosequencing datasets <i>(submitted)</i>
        </p>

      <footer class="site-footer">
        <span class="site-footer-owner">
            <a href="https://github.com/yana-safonova/ig_repertoire_constructor">IgReC</a> is maintained by
            Center for Algorithmic Biotechnology, Saint Petersburg University
        </span>
      </footer>

    </section>

  </body>
</html>

