<head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
    <title>igQUAST Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
    <style>
    </style>
</head>

<h1>igQUAST manual</h1>
1. <a href="#intro">What is igQUAST?</a><br>

2. <a href="#install">Installation</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.1. <a href="#test_datasets">Verifying your installation</a><br>

3. <a href="#igquast_usage">igQUAST usage</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#igrec_basic">Basic options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#igrec_advanced">Advanced options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#igrec_examples">Examples</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.4. <a href="#igrec_output">Output files</a><br>

4. <a href = "#examples">Examples</a><br>

5.  <a href="#repertoire_files">Antibody repertoire representation</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;6.1. <a href="#clusters_fasta">CLUSTERS.FASTA file format</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;6.2. <a href="#read_cluster_map">RCM file format</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;6.3. <a href="#alignment_info">Alignment Info file format</a><br>

6. <a href="#feedback">Feedback and bug reports</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;7.1. <a href="#citation">Citation</a><br>

<!--- ----------------- -->

<a id="intro"></a>
<h2>1. What is igQUAST?</h2>
igQUAST (ImmunoGlobulin QUality ASsessment Tool) is a tool for antibody repertoires quality assessment.
igQUAST also could be used for T-cells repertoires.
igQUAST can be used for benchmarking of antibody repertoire construction tools and for quality estimation of constructed repertoires.
igQUAST provides reference-based and reference-free analysis:
<ul>
    <li>During reference-based analysis the tool computes sensitivity and
        precision of reconstructed repertoire, detects undercorrected
        (erroneously split) and overcorrected (erroneously joined) clusters
        and estimates rate of correctly fixed initial reads.
    </li>
    <li>During reference-free analysis the tool estimates error rate and error profile of the initial RepSeq library. Also it detects overestimated clusters using special probabilistic model.
    </li>
</ul>

<a id="install"></a>
<h2>2. Installation</h2>

    <h2 id=install>2. Installation</h2>
aimQUAST has the following dependencies:
<ul>
    <li>64-bit Linux system</li>
    <li>g++ (version 4.7 or higher)</li>
    <li>cmake (version 2.8.8 or higher)</li>
    <li>Python 2 (version 2.7 or higher), including:
    <ul>
        <li><a href = "http://biopython.org/wiki/Download">BioPython</a></li>
        <li><a href = "http://matplotlib.org/users/installing.html">Matplotlib</a></li>
        <li><a href = "http://www.numpy.org/">NumPy</a></li>
        <li><a href = "http://www.scipy.org/install.html">SciPy</a></li>
        <li><a href = "http://pandas.pydata.org/">pandas</a></li>
        <li><a href = "https://web.stanford.edu/~mwaskom/software/seaborn/">Seaborn</a></li>
    </ul></li>
</ul>

	aimQUAST is a part of IgRepertoireConstructor package.<br> See <a
		href="manual.html#install">IgRepertoireConstructor manual</a> for
	installation instructions.
	<br>
	<br>
	Please verify your aimQUAST installation
	prior to initiate the aimQUAST:
    <pre class="code">
<code>
    ./aimquast.py --test
</code>
</pre> If the installation is successful, you will find the following information at the end of the log:

    <pre class="code">
<code>
  Thank you for using aimQUAST!
  Log was written to aimquast_test/aimquast.log
</code>
</pre>

<h2 id="usage">3. aimQUAST usage</h2>
    To run aimQUAST, type:
    <pre class="code">
    <code>
    ./aimQUAST.py [options] -s &lt;initial reads&gt; -c &lt;constructed repertoire FASTA&gt; -C &lt;constructed repertoire RCM&gt; -r &lt;reference repertoire FASTA&gt; -R &lt;reference repertoire RCM&gt; -o &lt;output dir for plots&gt;
    </code>
</pre>

    <!-- -->
    <h3 id="aimquastbasic">3.1. Basic options</h3>

    <code>-c / --constructed-repertoire &lt;constructed repertoire FASTA&gt;</code><br> FASTA file with constructed repertoire sequences. Could be gzipped.
    <br><br>

    <code>-C / --constructed-rcm &lt;constructed repertoire RCM&gt;</code><br> RCM file with constructed repertoire read-cluster map. Could be gzipped.
    <br><br>

    <code>-r / --reference-repertoire &lt;reference repertoire FASTA&gt;</code><br> FASTA file with reference repertoire sequences. Could be gzipped.
    <br><br>

    <code>-R / --reference-rcm &lt;reference repertoire RCM&gt;</code><br> RCM file with reference repertoire read-cluster map. Could be gzipped.
    <br><br>

    <code>-s / --initial-reads &lt;initial reads&gt;</code><br> Initial RepSeq reads in FASTA or FASTQ format. Could be gzipped.
    <br><br>

    <code>-o / --output-dir &lt;output dir&gt;</code><br> output directory (required).
    <br><br>

    <code>--text &lt;text report file&gt;</code><br> File for text report output.
    <br><br>

    <code>--json &lt;JSON report file&gt;</code><br> File for JSON report output.
    <br><br>

    <code>--test</code><br> Running on the toy test dataset. Command line corresponding to the test run is equivalent to the following:
    <pre class="code">
    <code>
    ./aimquast.py -s aimquast_test_dataset/merged_reads.fq -o aimquast_test -c aimquast_test_dataset/igrec_bad/final_repertoire.fa -C aimquast_test_dataset/igrec_bad/final_repertoire.rcm -r aimquast_test_dataset/ideal_final_repertoire.fa -R aimquast_test_dataset/ideal_final_repertoire.rcm --text aimquast_test/aimquast.txt --json aimquast_test/aimquast.json
    </code>
</pre>

    <code>--help</code><br> Printing help.
    <br><br>

    <h3 id="aimquast_advanced">3.2. Advanced options</h3>

    <code>-F / --figure-format &lt;figure formats(s)&gt;</code><br> Figure format(s) for outputted plots. Allowed values are <code>png</code>, <code>pdf</code> and <code>svg</code>. Several values should be separated by commas. Use empty string for non-producing plots. Default value is <code>png</code>.
    <br><br>

    <code>--reference-size-cutoff &lt;positive integer&gt;</code><br> Cutoff for reference cluster size. Lesser reference clusters are discarded during repertoire-to-repertoire comparison. See related paper for comprehensive discussion. Default value is <code>5</code>.
    <br><br>

    <code>--export-bad-clusters</code><br> Export bad clusters during reference-free analysis.
    <!-- -->

    <h3 id="aimquast_examples">3.3. Examples</h3> Provide reference-free analysis only:
    <pre class="code">
    <code>
    ./aimquast.py -s aimquast_test_dataset/merged_reads.fq -o aimquast_test -c aimquast_test_dataset/igrec_good/final_repertoire.fa -C aimquast_test_dataset/igrec_bad/final_repertoire.rcm
    </code>
</pre> Do not plot figures, make JSON report only:
    <pre class="code">
    <code>
    ./aimquast.py -s aimquast_test_dataset/merged_reads.fq -o aimquast_test -c aimquast_test_dataset/igrec_good/final_repertoire.fa -C aimquast_test_dataset/igrec_bad/final_repertoire.rcm -r aimquast_test_dataset/ideal_final_repertoire.fa -R aimquast_test_dataset/ideal_final_repertoire.rcm --figure-format= --json aimquast_test/aimquast.json
    </code>
</pre>

    <h3 id="aimquast_output_files">3.4. Output files</h3> aimQUAST creates working directory (which name was specified using option <code>-o</code>) and outputs the following files there:

    <ul>
        <li>
            <b>aimquast.log</b> &mdash; full log file of aimQUAST
        </li>
        <li>
            <b>constructed.rcm, reference.rcm, constructed.fa.gz. reference.fa.gz</b> &mdash; Reconstructed missing repertoire files
        </li>
        <li>
			<b>reference_based</b> &mdash; Directory with reference-based plots:
			<ul>
				<li><b>min_cluster_size_choose</b> &mdash; sensitivity and precision depending on cluster size threshold for the analyzed repertoire</li>
				<li><b>octoplot</b> &mdash; sensitivity and precision distance distribution depending on cluster size threshold for the analyzed repertoire</li>
				<li><b>multiplicity_distribution</b> &mdash; cluster size distribution for analyzed and reference repertoires</li>
				<li><b>constructed_majority_secondary</b> &mdash; dependence of secondary votes count on majority votes count in the analyzed repertoire. This plot helps to detect overcorrection</li>
				<li><b>constructed_purity_distribution</b> &mdash; distribution of cluster purity in the analyzed repertoire. This plot helps to detect overcorrection</li>
				<li><b>reference_majority_secondary</b> &mdash; dependence of secondary votes count on majority votes count in the reference repertoire. This plot helps to detect undercorrection</li>
				<li><b>reference_purity_distribution</b> &mdash; distribution of cluster purity in the reference repertoire. This plot helps to detect undercorrection</li>
			</ul>
        </li>
        <li>
            <b>reference_free</b> &mdash; Directory with reference-free plots:
			<ul>
				<li><b>{constructed,reference}_cluster_error_profile</b> &mdash; error profile estimation for analyzed and reference repertoire respectively</li>
				<li><b>{constructed,reference}_distribution_of_errors_in_reads</b> &mdash; distribution of the number of errors in reads for analyzed and reference repertoire respectively</li>
				<li><b>{constructed,reference}_estimation_of_max_error_distribution</b> &mdash; for each cluster, max errors by position for analyzed and reference repertoire respectively. This plot helps to detect overcorrection in the simplest amplification-free model</li>
			</ul>
        </li>
        <li>
            <b>reference_free/bad_constructed_clusters</b> &mdash; Directory with clusters from analyzed repertoire,
			which are identified as overcorrected during reference-free analysis
        </li>
        <li>
            <b>reference_free/bad_reference_clusters</b> &mdash; Directory with clusters from reference repertoire,
			which are identified as overcorrected during reference-free analysis
        </li>
    </ul>

    Files for reports (in text and JSON) are specified by corresponding options.

    <!-- -------- -->
    <h2 id="feedback">4. Feedback and bug reports</h2> Your comments, bug reports, and suggestions are very welcome. They will help us to further improve aimQUAST.
    <br><br> If you have any trouble running aimQUAST, please send us the log file from the output directory.
    <br><br> Address for communications: <a href="mailto:igtools_support@googlegroups.com">igtools_support@googlegroups.com</a>.

    <!-- <h3 id="citation">4.1. Citation</h3> If you use aimQUAST in your research, please refer to TODO Update citation -->
    <!-- <a href="http://bioginformatics.oxfordjournals.org/content/31/12/i53.long" target="_blank">Safonova et al., 2015</a>. -->
</body>
