<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>IgQUAST Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
            white-space: pre-wrap;       /* css-3 */
            white-space: -moz-pre-wrap;  /* Mozilla, since 1999 */
            white-space: -pre-wrap;      /* Opera 4-6 */
            white-space: -o-pre-wrap;    /* Opera 7 */
            word-wrap: break-word;       /* Internet Explorer 5.5+ */
        }
    </style>
    <style>

    </style>
</head>

<body>
<h1>IgQUAST manual</h1>
1. <a href="#intro">What is IgQUAST?</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.1. <a href="#igquast_input">Input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;1.2. <a href="#igquast_output">Output</a><br>
2. <a href="#install">Installation</a><br>
3. <a href="#igquast_usage">IgQUAST usage</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#igquast_basic">Basic options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#igquast_advanced">Advanced options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#igquast_examples">Examples</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.4. <a href="#igquast_output_files">Output files</a><br>
4. <a href="#references">Citations</a><br>
5. <a href="#feedback">Feedback and bug reports</a><br>

    <!-- -------- -->

    <h2 id="intro">1. What is IgQUAST?</h2> IgQUAST (<b>I</b>mmuno<b>G</b>lobulin <b>QU</b>ality <b>AS</b>sessment <b>T</b>ool)
    is a tool for adaptive immune repertoires quality assessment.
    IgQUAST can be used for benchmarking of adaptive immune repertoire construction tools and for quality estimation of constructed repertoires.
    IgQUAST provides reference-based and reference-free analysis:
    <ul>
        <li>During reference-based analysis the tool computes sensitivity and
            precision of reconstructed repertoire, detects undercorrected
            (erroneously split) and overcorrected (erroneously joined) clusters
            and estimates rate of correctly fixed initial reads.
        </li>
        <li>During reference-free analysis the tool estimates error rate and error profile of the initial Rep-Seq library.
            <!-- Also it detects overestimated clusters using amplification&#45;free Poisson model. -->
        </li>
    </ul>

    <h3 id="igquast_input">1.1. Input</h3> IgQUAST takes as an input:
    <ul>
        <li>Initial Rep-Seq library;</li>
        <li>Analyzed adaptive immune repertoire constructed on this library;</li>
        <li>For reference-based analysis, reference repertoire constructed on the same library.</li>
    </ul>
    Initial Rep-Seq library should be in FASTA or FASTQ format.
        Reads should be properly cropped (should start from V gene beginning
        and finish by J gene ending) and reads from negative strand should
        be reversed.
        Cropping and strand correction of Rep-Seq reads may be performed by
        VJFinder tool from IgRepertoireConstructor package
        (see <a href="manual.html">IgRepertoireConstructor manual</a> for VJFinder input-output format description).
    <br>
    Analyzed and reference repertoires should be presented by two files each:
        repertoire sequences in FASTA format and read-cluster map (RCM)
        file in a special format. See <a href="manual.html#repertoire_files">IgRepertoireConstructor manual</a>
        for a comprehensive repertoire format description. One of these two
        files could be missing, in this case the tool reconstructs it using
        available information.


    <h3 id="igquast_output">1.2. Output</h3> IgQUAST reports:
    <ul>
        <li>Plots for visual analysis;</li>
        <li>Metrics report.</li>
    </ul>

    Plots are reported in PNG, PDF and SVG formats. Metrics are reported in text (brief) and JSON (full) formats.

    <h2 id=install>2. Installation</h2>
IgQUAST has the following dependencies:
<ul>
    <li>64-bit Linux system</li>
    <li>g++ (version 4.7 or higher)</li>
    <li>cmake (version 2.8.8 or higher)</li>
    <li>Python 2 (version 2.7 or higher), including:
    <ul>
        <li><a href = "http://biopython.org/wiki/Download">BioPython</a></li>
        <li><a href = "http://matplotlib.org/users/installing.html">Matplotlib</a></li>
        <li><a href = "http://www.numpy.org/">NumPy</a></li>
        <li><a href = "http://www.scipy.org/install.html">SciPy</a></li>
        <li><a href = "http://pandas.pydata.org/">pandas</a></li>
        <li><a href = "https://web.stanford.edu/~mwaskom/software/seaborn/">Seaborn</a></li>
    </ul></li>
</ul>

    IgQUAST is a part of IgRepertoireConstructor package.<br> See <a
        href="manual.html#install">IgRepertoireConstructor manual</a> for
    installation instructions.
    <br>
    <br>
    Please verify your IgQUAST installation
    prior to initiate the IgQUAST:
    <pre class="code">
<code>
    ./igquast.py --test
</code>
</pre> If the installation is successful, you will find the following information at the end of the log:

    <pre class="code">
<code>
  Thank you for using IgQUAST!
  Log was written to igquast_test/igquast.log
</code>
</pre>

<h2 id="usage">3. IgQUAST usage</h2>
    To run IgQUAST, type:
    <pre class="code">
    <code>
    ./igquast.py [options] -s &lt;initial reads&gt; -c &lt;constructed repertoire FASTA&gt; -C &lt;constructed repertoire RCM&gt; -r &lt;reference repertoire FASTA&gt; -R &lt;reference repertoire RCM&gt; -o &lt;output dir for plots&gt;
    </code>
</pre>

    <!-- -->
    <h3 id="igquastbasic">3.1. Basic options</h3>

    <code>-c / --constructed-repertoire &lt;constructed repertoire FASTA&gt;</code><br> FASTA file with constructed repertoire sequences. Could be gzipped.
    <br><br>

    <code>-C / --constructed-rcm &lt;constructed repertoire RCM&gt;</code><br> RCM file with constructed repertoire read-cluster map. Could be gzipped.
    <br><br>

    <code>-r / --reference-repertoire &lt;reference repertoire FASTA&gt;</code><br> FASTA file with reference repertoire sequences. Could be gzipped.
    <br><br>

    <code>-R / --reference-rcm &lt;reference repertoire RCM&gt;</code><br> RCM file with reference repertoire read-cluster map. Could be gzipped.
    <br><br>

    <code>-s / --initial-reads &lt;initial reads&gt;</code><br> Initial Rep-Seq reads in FASTA or FASTQ format. Could be gzipped.
    <br><br>

    <code>-o / --output-dir &lt;output dir&gt;</code><br> output directory (required).
    <br><br>

    <code>--text &lt;text report file&gt;</code><br> File for text report output.
    Default: <code>&lt;output dir&gt;/igquast.txt</code>
    <br><br>

    <code>--json &lt;JSON report file&gt;</code><br> File for JSON report output.
    Default: <code>&lt;output dir&gt;/igquast.json</code>
    <br><br>

    <code>--test</code><br> Running on the toy test dataset. Command line corresponding to the test run is equivalent to the following:
    <pre class="code">
    <code>
    ./igquast.py -s igquast_test_dataset/merged_reads.fq -o igquast_test -c igquast_test_dataset/igrec_bad/final_repertoire.fa -C igquast_test_dataset/igrec_bad/final_repertoire.rcm -r igquast_test_dataset/ideal_final_repertoire.fa -R igquast_test_dataset/ideal_final_repertoire.rcm --text igquast_test/igquast.txt --json igquast_test/igquast.json
    </code>
</pre>

    <code>-h / --help</code><br> Printing help.
    <br><br>

    <h3 id="igquast_advanced">3.2. Advanced options</h3>

    <code>-F / --figure-format &lt;figure formats(s)&gt;</code><br> Figure
    format(s) for outputted plots. Allowed values are <code>png</code>,
    <code>pdf</code> and <code>svg</code>. Several values should be separated
    by commas. Use empty string for non-producing plots. Default value is
    <code>png</code>.
    <br><br>

    <code>--tau &lt;positive integer&gt;</code><br>Maximal distance for repertoire-to-repertoire matching (default: 6)
    <br><br>

    <code>--reference-size-cutoff &lt;positive integer&gt;</code><br> Cutoff
    for reference cluster size. Lesser reference clusters are discarded during
    repertoire-to-repertoire comparison. See related paper for comprehensive
    discussion. Default value is <code>5</code>.
    <br><br>

    <code>--export-bad-clusters</code><br> Export untrustworthy clusters during reference-free analysis.
    <br><br>
    <code>--experimental</code><br> Enable experimental features.
    <br><br>
    <code>--rcm-based</code><br> Enable RCM-based metrics and plots. They are disabled by default since their computation is very time- and
    memory-consuming.
    <br><br>



    <h3 id="igquast_examples">3.3. Examples</h3> Provide reference-free analysis only:
    <pre class="code">
    <code>
    ./igquast.py -s igquast_test_dataset/merged_reads.fq -o igquast_test -c igquast_test_dataset/igrec_good/final_repertoire.fa -C igquast_test_dataset/igrec_bad/final_repertoire.rcm
    </code>
</pre> Do not plot figures, make JSON report only:
    <pre class="code">
    <code>
    ./igquast.py -s igquast_test_dataset/merged_reads.fq -o igquast_test -c igquast_test_dataset/igrec_good/final_repertoire.fa -C igquast_test_dataset/igrec_bad/final_repertoire.rcm -r igquast_test_dataset/ideal_final_repertoire.fa -R igquast_test_dataset/ideal_final_repertoire.rcm --figure-format=
    </code>
</pre>

    <h3 id="igquast_output_files">3.4. Output files</h3> IgQUAST creates working directory
    (which name is specified using option <code>-o</code>) and outputs the following files there:

    <ul>
        <li>
            <b>igquast.log</b> &mdash; full log of the IgQUAST run
        </li>
        <li>
            <b>constructed.rcm, reference.rcm, constructed.fa.gz. reference.fa.gz</b> &mdash; Reconstructed missing repertoire files
        </li>
        <li>
            <b>reference_based</b> &mdash; Directory with reference-based plots:
            <ul>
                <li><b>error_pos_dist</b> &mdash; TBD</li>
                <li><b>min_cluster_size_choose</b> &mdash; sensitivity and precision depending on cluster size threshold for the analyzed repertoire</li>
                <li><b>octoplot</b> &mdash; constructed to reference and reference to constructed distance distribution depending on cluster size threshold for the analyzed repertoire. 8 plots in all on one figure</li>
                <li><b>{constructed_to_reference,reference_to_constructed}_distance_distribution_size_{1,3,5,10}</b> &mdash; the same 8 plots separately</li>
                <li><b>multiplicity_distribution</b> &mdash; cluster size distribution for analyzed and reference repertoires</li>

                <li><b>constructed_majority_secondary</b> &mdash; dependence of secondary votes count on majority votes count in the analyzed repertoire. This plot helps to detect overcorrection</li>
                <li><b>constructed_majority_secondary_large</b> &mdash; the same plot for large clusters only</li>
                <li><b>constructed_majority_secondary_large_ylog</b> &mdash; the same plot with logarithmic Y-scale</li>

                <li><b>constructed_purity_distribution</b> &mdash; distribution of cluster purity in the analyzed repertoire. This plot helps to detect overcorrection</li>
                <li><b>constructed_purity_distribution_large</b> &mdash; the same plot for large clusters only</li>
                <li><b>constructed_purity_distribution_large_ylog</b> &mdash; the same plot with logarithmic Y-scale</li>

                <li><b>reference_majority_secondary</b> &mdash; dependence of secondary votes count on majority votes count in the reference repertoire. This plot helps to detect undercorrection</li>
                <li><b>reference_majority_secondary_large</b> &mdash; the same plot for large clusters only</li>
                <li><b>reference_majority_secondary_large_ylog</b> &mdash; the same plot with logarithmic Y-scale</li>

                <li><b>reference_purity_distribution</b> &mdash; distribution of cluster purity in the reference repertoire. This plot helps to detect undercorrection</li>
                <li><b>reference_purity_distribution_large</b> &mdash; the same plot fro large clusters only</li>
                <li><b>reference_purity_distribution_large_ylog</b> &mdash; the same plot with logarithmic Y-scale</li>

                <li><b>reference_discordance_distribution</b> &mdash; TBD</li>
                <li><b>constructed_discordance_distribution</b> &mdash; TBD</li>
            </ul>
        </li>
        <li>
            <b>reference_free</b> &mdash; Directory with reference-free plots:
            <ul>
                <li><b>{constructed,reference}_cluster_error_profile</b> &mdash; error profile estimation for analyzed and reference repertoire respectively</li>
                <li><b>{constructed,reference}_distribution_of_errors_in_reads</b> &mdash; distribution of the number of errors in reads for analyzed and reference repertoire respectively</li>
                <li><b>{constructed,reference}_estimation_of_max_error_distribution</b> &mdash; for each cluster, max errors by position for analyzed and reference repertoire respectively. This plot helps to detect overcorrection in the simplest amplification-free model</li>
                <li>
                    <b>bad_constructed_clusters</b> &mdash; Directory with clusters from analyzed repertoire,
                    which are identified as overcorrected during reference-free analysis
                </li>
                <li>
                    <b>bad_reference_clusters</b> &mdash; Directory with clusters from reference repertoire,
                    which are identified as overcorrected during reference-free analysis
                </li>
            </ul>
        </li>
    </ul>

    Files for the reports (in text and JSON formats) are specified by the corresponding options.

    <h2 id = "references">4. Citations</h2>
    <p align = "justify">
        Alexander Shlemov, Sergey Bankevich, Andrey Bzikadze,
        Dmitriy Chudakov,
        Yana Safonova,
        and Pavel A. Pevzner.
        Reconstructing antibody repertoires from error-prone immunosequencing datasets <i>(submitted to Genome Research)</i>
    </p>
    <!-- -------- -->
    <h2 id="feedback">5. Feedback and bug reports</h2> Your comments, bug
    reports, and suggestions are very welcome. They will help us to further
    improve IgQUAST.
    <br><br> If you have any trouble running IgQUAST, please send us the log file from the output directory.
    <br><br> Address for communications: <a href="mailto:igtools_support@googlegroups.com">igtools_support@googlegroups.com</a>.
</body>
