<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>IgRepertoireConstructor by yana-safonova</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name" align=center>IgRepertoireConstructor</h1>
      <h2 class="project-tagline" align=center>An algorithm for error-correction of immunosequencing Illumina MiSeq reads and immunoproteogenomics analysis</h2>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor" class="btn"><b>View on GitHub</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/zipball/master" class="btn"><b>Download .zip</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/tarball/master" class="btn"><b>Download .tar.gz</b></a>
    </section>

    <section class="main-content">
    <h2>Content</h2>
    <p align = "center">
        <ul>
            <li><a href = "#motivation">Motivation</a></br></li>
            <li><a href = "#approach">Our approach</a></li>
            <li><a href = "#ig_repertoire_constructor">IgRepertoireConstructor</a></li>
            <ul>
                <li><a href = "#vj_finder">VJ Finder</a></li>
                <li><a href = "#graph_algos">Construction and clusterizarion of Smith-Waterman graph</a></li>
                <li><a href = "#antibody_construction">Antibody Construction</a></li>
            </ul>
            <li><a href = "#mass_spectra_analysis">Mass Spectra Analyzer</a></li>
            <li><a href = "#dsf">DSF (Dense Subgraph Finder)</a></li>
            <li><a href = "#references">Manual and citations</a></li>
        </ul>
     </p>

    <h2 id = "motivation">Motivation</h2>
    <p align = "justify">
        Construction of antibody repertoire is a very preliminary step of any immunological analysis based on immunosequencing reads.
        Accurate construction of antibody repertoire avoids erroneous analysis of natural variations and antibody abundances.
    </p>

    <h2 id = "approach">Our approach</h2>
    <p align = "justify>
        <b>IgRepertoireConstructor</b> constructs antibody repertoire from immunosequencing reads and
        uses mass spectra to validate the constructed antibody repertoires.
        IgRepertoireConstructor pipeline includes the following steps:
    <ul>
            <li> <b>IgRepertoireConstructor:</b> construction of antibody repertoire from immunosequencing reads.
                IgRepertoireConstructor corrects sequencing and amplification errors and
                joins together reads corresponding to the identical antibodies.
                Constructed repertoire can be used as an highly accurate input for various immunological problems, e.g.:
                <ul>
                    <li><i>Construction of clonal trees.</i> TODO: add description here.</li>
                    <li><i>V(D)J labeling / CDR3 classification.</i> TODO: add description here.</li>
                    <li><i>Mass spectra identification.</i> Constructed repertoire can be used as a database for MS search.</li>
                </ul>
            <li> <b>MassSpectraAnalyzer:</b> analysis of results of MS matching.</li>
    </ul>
    <table align = center>
    <tr><td align = center><img src="figures/pipeline.png" width = "80%" alt="pipeline"></td></tr>
    </table>

    <h2 id = "ig_repertoire_constructor">IgRepertoireConstructor</h2>
    <p>
            IgRepertoireConstructor takes as an input reads covering variable region of antibodies and
            corrects sequencing and amplification errors.
            Algorithm performs the following steps:
            <ul>
                <li><b>VJ Finder.</b> TODO: add descrpition and examples of output</li>
                <li><b>Construction & clusterization of SW graph.</b> TODO: add description.</li>
                <li><b>Antibody construction.</b> TODO: add description here</li>
            </ul>
        <h3 id = "vj_finder">VJ Finder</h3>
        TODO: add some description here

        <h3 id = "graph_algos">Construction and clusterizarion of Smith-Waterman graph</h3>
        <table align = center border = "0">
        <tr>
            <td align = center><img src="figures/graph_357134_default.jpg" alt = "hamming_graph_before_clusterization" width = 300px ></td>
            <td align = center><img src="figures/graph_357134_colored.jpg" alt = "hamming_graph_after_clusterization" width = 300px></td> 
        </tr>
        <tr>
            <td><small>Example of connected component of Hamming graph constructed from immunosequencing reads (human, heavy chain repertoire of IgM).</small></td>
            <td><small>Clusterization of Hamming graph performed by IgRepertoireConstructor. Each color corresponds to single antibody.</small></td>
        </tr>
        </table></br>

        <h3 id = "antibody_construction">Antibody Construction</h3>
        TODO: add description

    <h2 id = "mass_spectra_analysis">Mass Spectra Analyzer</h2>
    <p align = "justify">
        This step performs immunoproteogenomics analysis to validate constructed repertoire using mass spectra.
        It takes alignment of mass spectra against constructed repertoire in
        <a href = "http://www.psidev.info/mzidentml">mzIdentML 1.1 format</a>
        (e.g., generated by MS-GF+) as an input and outputs a set of metrics and graphs that show
        how close are the constructed repertoire and provided mass spectra:

        <table align = center border = "0">
        <tr>
            <td align = center><img src="figures/ms_analysis/peptide_length_Total.png" alt = "peptide_length_distribution"></td>
            <td align = center><img src="figures/ms_analysis/PSM_cov_Total.png" alt = "psm_coverage_histogram"></td> 
        </tr>
        <tr>
            <td align = center><small>Histogram of peptide length distribution.</small></td>
            <td><small>Histogram of variable region coverage by PSMs. Grey bars corresponds to expected positions of CDRs.</small></td>
        </tr>
        </table></br>
    </p>

        <h2 id="dsf">DSF (Dense Subgraph Finder)</h2>
        <p align = "justify">
        Search for <i>dense subgraphs</i> (or <i>corrupted cliques</i>) is a very common problem arising in bioinformatics
        (e.g., co-expression of genes) and studies of social interactions (e.g., recommendation services).
        We decided to make dense subgraph search available as an independent tool.
        One can download IgRepertoireConstructor archive and run DSF algorithm for search of dense subgraph.</br></br>

        DSF takes as an input graph in GRAPH format (please find specification <a href = "">here</a>) and outputs
        decomposition where the identical ids were assigned for vertices from the same dense subgraph.
        Density of constructed subgraph can be tuned via parameter <code>-f</code> that means minimum edge fill-in
        in dense subgraphs. Table below shows as dense subgraph decomposition varies dependending of value of minimal edge fill-in.

        <table align = center border="0">
            <tr>
                <td align = center>
                    <div style="width: 380px">
                        <img src = "docs/dsf_0.3.png" alt = "dsf_example_edge_fill-in_0.3" width = 270px>
                    </div>
                </td>
                <td align = center>
                    <div style="width: 380px">
                        <img src = "docs/dsf_0.4.png" alt = "dsf_example_edge_fill-in_0.4" width = 270px>
                    </div>
                </td>
            </tr>
            <tr>
                <td align = center><b><small>Minimal edge fill-in = 0.3</small></b></td>
                <td align = center><b><small>Minimal edge fill-in = 0.4</small></b></td>
            </tr>
            <tr>
                <td align = center>
                    <small>
                    # dense subgraphs = 55</br>
                    # trivial subgraphs = 40</br>
                    size of maximal subgraph = 157
                    </small>
                </td>
                <td align = center>
                    <small>
                    # dense subgraphs = 61</br>
                    # trivial subgraphs = 46</br>
                    size of maximal subgraph = 104
                    </small>
                </td>
            </tr>
            <tr>
                <td align = center>
                    <img src = "docs/dsf_0.5.png" alt = "dsf_example_edge_fill-in_0.5" width = 270px>
                </td>
                <td align = center>
                    <img src = "docs/dsf_0.7.png" alt = "dsf_example_edge_fill-in_0.7" width = 270px>
                </td>
            </tr>
            <tr>
                <td align = center><b><small>Minimal edge fill-in = 0.5</small></b></td>
                <td align = center><b><small>Minimal edge fill-in = 0.7</small></b></td>
            </tr>
            <tr>
                <td align = center>
                    <small>
                    # dense subgraphs = 60</br>
                    # trivial subgraphs = 43</br>
                    size of maximal subgraph = 87
                    </small>
                </td>
                <td align = center>
                    <small>
                    # dense subgraphs = 68</br>
                    # trivial subgraphs = 42</br>
                    size of maximal subgraph = 43
                    </small>
                </td>
            </tr>
         </table>
    </p></br></br>


    <h2 id = "references">Manual and citations</h2>
    <p align = "center">
        <ul>
            <li>IgRepertoireConstructor manual can be found <a href = 'ig_repertoire_constructor_manual.html'>here</a>.</li>
            <li>DSF manual can be found <a href = 'dsf_manual.html'>here</a>.</li>
        </ul>
    </p>

    <p align = "justify">
        <b>Citation:</b> Yana Safonova, Stefano Bonissone, Eugene Kurpilyansky, Ekaterina Starostina, Alla Lapidus,
        Jeremy Stinson, Laura DePalatis, Wendy Sandoval, Jennie Lill and Pavel A. Pevzner.
        IgRepertoireConstructor: a novel algorithm for antibody repertoire construction and immunoproteogenomics analysis.
        Bioinformatics. 2015 Jun 15; 31(12): i53-i61. doi: 10.1093/bioinformatics/btv238.
    </p>

      <footer class="site-footer">
            <p align = center>
                <big>Possibly you also will be interested in other our immunogenomics tools:</br> <a href = 'http://yana-safonova.github.io/ig_simulator/'><b>IgSimulator, a versatile immunosequencing simulator</b></a>.<big>
            </p>            
      </footer>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/yana-safonova/ig_simulator">IgRepertoireConstructor</a> is maintained by <a href="https://github.com/yana-safonova">Yana Safonova</a> (Center for Algorithmic Biotechnology, Saint Petersburg State University).</span>
      </footer>

    </section>

  
  </body>
</html>

