<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>IgRepertoireConstructor by yana-safonova</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name" align=center>IgRepertoireConstructor</h1>
      <h2 class="project-tagline" align=center>An algorithm for error-correction of immunosequencing Illumina MiSeq reads and immunoproteogenomics analysis</h2>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor" class="btn"><b>View on GitHub</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/zipball/master" class="btn"><b>Download .zip</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/tarball/master" class="btn"><b>Download .tar.gz</b></a>
    </section>

    <section class="main-content">
    
    <p><b>IgRepertoireConstructor</b> performs error-correction of immunosequencing reads and uses mass spectra to validate the constructed antibody repertoires. IgRepertoireConstructor pipeline includes the following steps:
    <ul>
            <li> <b>IgDataCleaner:</b> merging paired-end Illumina MiSeq reads and cleaning contaminations.</li>
            <li> <b>IgRepertoireConstructor:</b> construction of antibody repertoire from cleaned reads. Constructed antibody repertoire can be used as a database for matching of mass spectra, e.g., using MS-GF+ tool.</li> 
            <li> <b>MassSpectraAnalyzer:</b> analysis of results of MS matching.</li>
    </ul>
    <table align = center>
    <tr><td align = center><img src="figures/pipeline.png" width = "80%" alt="pipeline"></td></tr>
    </table></br>

    <ul>
        <li><b>IgDataCleaner</b>. 
            IgDataCleaner takes Illumina MiSeq paired end reads as an input and expect that each left and right reads overlap and cover variable region of antibody. 
            IgDataCleaner stitches paired-end reads and removes contaminations using alignment of stitched reads against Ig germline database. 
            This procedure significantly improves quality of input reads (left) and allows one to constructs reads covering variable region of antibody (right):</li>
        <table align = center border = "0">
        <tr> 
            <td align = center><img src="figures/dataset_preparation/7_SAM15574987_aver_quality.png" alt = "read_quality_comparison"></td> 
            <td align = center><img src="figures/dataset_preparation/7_SAM15574987_merged_read_length.png" alt = "merged_read_length"></td>
        </tr>
        <tr>
            <td><small>Average quality of reads before (red) and after (blue) merging</small></td>
            <td><small>Histogram of merged read length distribution (human, heavy chain repertoire of IgM)</small></td>
        </tr>
        </table></br>

        <li><b>IgRepertoireConstructor.</b> 
            IgRepertoireConstructor takes as an input reads covering variable region of antibodies and corrects Illumina sequencing errors. 
            Algorithm constructs Hamming graph on input reads and uses accurate clustering techniques for reconstruction of original antibody clusters:</li>

        <table align = center border = "0">
        <tr>
            <td align = center><img src="figures/graph_357134_default.jpg" alt = "hamming_graph_before_clusterization" width = 300px ></td>
            <td align = center><img src="figures/graph_357134_colored.jpg" alt = "hamming_graph_after_clusterization" width = 300px></td> 
        </tr>
        <tr>
            <td><small>Example of connected component of Hamming graph constructed from immunosequencing reads (human, heavy chain repertoire of IgM).</small></td>
            <td><small>Clusterization of Hamming graph performed by IgRepertoireConstructor. Each color corresponds to single antibody.</small></td>
        </tr>
        </table></br>

        <li><b>MassSpectraAnalyzer.</b> This step performs immunoproteogenomics analysis to validate constructed repertoire using mass spectra. 
            It takes alignment of mass spectra against constructed repertoire in <a href = "http://www.psidev.info/mzidentml">mzIdentML 1.1 format</a> (e.g., generated by MS-GF+) as an input and outputs a set of metrics and graphs that show how close are the constructed repertoire and provided mass spectra:</li>

        <table align = center border = "0">
        <tr>
            <td align = center><img src="figures/ms_analysis/peptide_length_Total.png" alt = "peptide_length_distribution"></td>
            <td align = center><img src="figures/ms_analysis/PSM_cov_Total.png" alt = "psm_coverage_histogram"></td> 
        </tr>
        <tr>
            <td align = center><small>Histogram of peptide length distribution.</small></td>
            <td><small>Histogram of variable region coverage by PSMs. Grey bars corresponds to expected positions of CDRs.</small></td>
        </tr>
        </table></br>
    </ul>
    </p>

<p align = "justify">
    IgRepertoireConstructor manual can be found <a href = 'ig_repertoire_constructor_manual.html'>here</a>.</br></br>

    <b>Citation:</b> Yana Safonova, Stefano Bonissone, Eugene Kurpilyansky, Ekaterina Starostina, Alla Lapidus, Jeremy Stinson, Laura DePalatis, Wendy Sandoval, Jennie Lill and Pavel A. Pevzner. IgRepertoireConstructor: a novel algorithm for antibody repertoire construction and immunoproteogenomics analysis. Bioinformatics. 2015 Jun 15;31(12):i53-i61. doi: 10.1093/bioinformatics/btv238.
</p>

      <footer class="site-footer">
            <p align = center>
                <big>You possibly also will be interested in other our immunogenomics tools:</br> <a href = 'http://yana-safonova.github.io/ig_simulator/'><b>IgSimulator, a versatile immunosequencing simulator</b></a>.<big>
            </p>            
      </footer>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/yana-safonova/ig_simulator">IgRepertoireConstructor</a> is maintained by <a href="https://github.com/yana-safonova">Yana Safonova</a> (Center for Algorithmic Biotechnology, Saint Petersburg State University).</span>

<!--        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span> 
-->
      </footer>

    </section>

  
  </body>
</html>

