<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>IgRepertoireConstructor by yana-safonova</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name" align=center>IgRepertoireConstructor</h1>
      <h2 class="project-tagline" align=center>An algorithm for error-correction of immunosequencing Illumina MiSeq reads and immunoproteogenomics analysis</h2>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor" class="btn"><b>View on GitHub</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/zipball/master" class="btn"><b>Download .zip</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/tarball/master" class="btn"><b>Download .tar.gz</b></a>
    </section>

    <section class="main-content">
    
    <p><b>IgRepertoireConstructor</b> performs error-correction of immunosequencing reads and uses mass spectra to validate the constructed antibody repertoires. IgRepertoireConstructor pipeline is presented below:</br>
    <table align = center>
    <tr><td align = center><img src="figures/pipeline.png" width = "80%" alt="pipeline"></td></tr>
    </table></br>

    <ul>
        <li><b>Library preparation</b>. 
            IgRepertoireConstructor takes Illumina MiSeq paired end reads as an input and expect that each left and right reads overlap and cover variable region of antibody. 
            IgRepertoireConstructor stitches paired-end reads and removes contaminations using alignment of stitched reads against Ig germline database. 
            This procedure significantly improves quality of input reads (left) and allows one to constructs reads covering variable region of antibody (right):</li>
        <table align = center border = "0">
        <tr> 
            <td align = center><img src="figures/dataset_preparation/7_SAM15574987_aver_quality.png" alt = "read_quality_comparison"></td> 
            <td align = center><img src="figures/dataset_preparation/7_SAM15574987_merged_read_length.png" alt = "merged_read_length"></td>
        </tr>
        <tr>
            <td><small>Average quality of reads before (red) and after (blue) merging</small></td>
            <td><small>Histogram of merged read length distribution (human, heavy chain repertoire of IgM)</small></td>
        </tr>
        </table></br>

        <li><b>Construction of antibody repertoire.</b> 
            IgRepertoireConstructor takes as an input reads covering variable region of antibodies and corrects Illumina sequencing errors. 
            Algorithm constructs Hamming graph on input reads and uses accurate clustering techniques for reconstruction of original antibody clusters:</li>

        <table align = center border = "0">
        <tr>
            <td align = center><img src="figures/graph_357134_default.jpg" alt = "hamming_graph_before_clusterization" width = 300px ></td>
            <td align = center><img src="figures/graph_357134_colored.jpg" alt = "hamming_graph_after_clusterization" width = 300px></td> 
        </tr>
        <tr>
            <td><small>Example of connected component of Hamming graph constructed from immunosequencing reads (human, heavy chain repertoire of IgM).</small></td>
            <td><small>Clusterization of Hamming graph performed by IgRepertoireConstructor. Each color corresponds to single antibody.</small></td>
        </tr>
        </table></br>

        <li><b>Immunoproteogenomics analysis.</b> Constructed repertoire can be used as database for peptide identification from mass spectra. IgRepertoireConstructor takes as an input MZID file and computes various statistics about constructed repertoire.</li>
    </ul>
    </p>

      <footer class="site-footer">
            <p align = center>
                <big>You possibly also will be interested in other our immunogenomics tools:</br> <a href = 'http://yana-safonova.github.io/ig_simulator/'><b>IgSimulator, a versatile immunosequencing simulator</b></a>.<big>
            </p>            
      </footer>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/yana-safonova/ig_simulator">IgRepertoireConstructor</a> is maintained by <a href="https://github.com/yana-safonova">Yana Safonova</a> (Center for Algorithmic Biotechnology, Saint Petersburg State University).</span>

<!--        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span> 
-->
      </footer>

    </section>

  
  </body>
</html>

