<!DOCTYPE html>

<script>

    var button_beg = '<button id="button" onclick="showhide()">', button_end = '</button>';
    var show_button = 'Show related works', hide_button = 'Hide';
    function showhide() {
        var div = document.getElementById( "hide_show" );
        var showhide = document.getElementById( "showhide" );
        if ( div.style.display !== "none" ) {
            div.style.display = "none";
            button = show_button;
            showhide.innerHTML = button_beg + button + button_end;
        } else {
            div.style.display = "block";
            button = hide_button;
            showhide.innerHTML = button_beg + button + button_end;
        }
    }
    function setup_button( status ) {
        if ( status == 'show' ) {
            button = hide_button;
        } else {
            button = show_button;
        }
        var showhide = document.getElementById( "showhide" );
        showhide.innerHTML = button_beg + button + button_end;
    }
    window.onload = function () {
        setup_button( 'hide' );
        showhide(); // if setup_button is set to 'show' comment this line
    }
</script>

<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Y-Tools</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name" align=center>Y-Tools</h1>
      <h2 class="project-tagline" align=center>tools for construction of full-length adaptive immune repertoires and
          immunoinformatics analysis</h2>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor" class="btn"><b>View on GitHub</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/zipball/master" class="btn"><b>Download .zip</b></a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/tarball/master" class="btn"><b>Download .tar.gz</b></a>
    </section>

    <section class="main-content">

        <h2 id = "motivation">Rep-seq or full-length sequencing of adaptive immune repertoires</h2>

        <p align = justify>
            Modern sequencing technologies (e.g., Illumina MiSeq) allow biologists to perform full-length scanning of
            adaptive immune repertoire.
            For example, a pair of overlapping paired-end Illumina MiSeq reads (250x2 or 300x2) is able to cover a variable region of antibody or TCR:
        </p>

        <div align = center>
            <img src = "docs/antibody_sequencing.png" width = 65% alt = "antibody_sequencing">
        </div>

        <!---
        <div id="showhide"></div>
        <div id="hide_show">
            <ul>
                <li>
                    Benichou J., Ben-Hamo R., Louzoun Y., Efroni S..
                    <a href = "https://www.ncbi.nlm.nih.gov/pubmed/22043864">Rep-Seq: uncovering the
                        immunological repertoire through next-generation sequencing.</a> <i>Immunology</i>. 2012. 135(3): pp. 183-91.
                </li>
                <li>
                    Turchaninova M.A., Davydov A., Britanova O.V., Shugay M., Bikos V., Egorov E.S.,
                    Kirgizova V.I., Merzlyak E.M., Staroverov D.B., Bolotin D.A., Mamedov I.Z., Izraelson M.,
                    Logacheva M.D., Kladova O., Plevova K., Pospisilova S., Chudakov D.M.
                    <a href = "http://www.nature.com/nprot/journal/v11/n9/abs/nprot.2016.093.html">High-quality
                        full-length immunoglobulin profiling with unique molecular barcoding.</a>
                    <i>Nat Protoc.</i> 2016. 11(9): pp. 1599-616.
                </li>
            </ul>
        </div>
        --->

        <h3>Our tools for analysis of adaptive immune repertoires</h3>
        <p align = justify>
            Sequencing data of adaptive immune repertoires or <i>Rep-seq</i> is an input of various immunological studies.
            We propose a set of tools for solving related immunoinformatics problems:
        <ul>
            <li>Repertoire construction and its quality assessment</li>
            <li>Immunoproteogenomics</li>
            <li>Diversity and clonal analysis</li>
        </ul>
        </p>


        <table align = "center" width = "100%">
            <tr>
                <td align = center width="1000000px">
                    <a href = "docs/immunogenomics_pipeline.svg">
                        <img src="docs/immunogenomics_pipeline.svg" width = "100%" alt="pipeline">
                    </a>
                </td>
            </tr>
            <tr>
                <td align = center><small>Y-tools in a nutshell</small></td>
            </tr>
        </table>

        <ul>
            <li>
                <a href = "#igrec_description"><b>IgReC</b></a> performs construction of antibody repertoire from immunosequencing reads:
                it corrects sequencing and amplification errors and joins together reads corresponding
                to identical antibodies.
                Constructed repertoire present a highly accurate input for various immunological applications, e.g.,
                construction of clonal trees, V(D)J labeling / CDR3 classification, mass spectra identification etc.
            </li><br>
            <li>
                If your sequencing data contains unique molecular identifiers (or UMIs),
                we recommend to use <a href = "barcodedIgReC.html"><b>Barcoded IgReC</b></a>,
                a modification of IgReC tool for processing barcoded Rep-seq libraries.
            </li><br>
            <li>
                <a href = "diversity_analyzer.html"><b>Diversity Analyzer:</b></a> a tool for diversity analysis of Rep-seq data.
                Diversity Analyzer computes VJ, CDR, and SHM labelings of input receptor sequences.
                It also visualizes various statistics that help to analyze characteristics of an input dataset.
            </li><br>
            <li>
                <a href = "igquast.html"><b>IgQUAST:</b></a> a quality assessment tool for full-length adaptive immune repertoires.
                IgQUAST analyzes a repertoire based on an original Rep-seq library and a reference repertoire (if available).
                IgQUAST detects over- and undercorrection of clusters, computes sensitivity and specificity of a repertoire,
                and analyzes repertoire abundances.
            </li><br>
            <li>
                <a href = "antevolo.html"><b>AntEvolo:</b></a> an algorithm for construction of clonal trees for an antibody repertoire.
            </li><br>
            <li>
                <a href = "#ms_analyzer_description"><b>Mass Spectra Analyzer</b></a> is a tool for immunoproteogenomics analysis.
                Repertoire constructed by IgReC can be converted into a
                database and used for mass spectra identifications (using some standard tool,
                e.g., <a href = "https://omics.pnl.gov/software/ms-gf">MS-GF+</a>).
                Mass Spectra Analyzer performs analysis of the computed matches and computes various statistics
                showing similarity of the constructed repertoire and mass spectra.
            </li>
        </ul>

        <hr>

        <!-------------------------------------------------------------- --->
        <h2>Construction of full-length adaptive immune repertoires</h2>
        <h3>Repertoire construction problem</h3>
        <p align = "justify">
            Repertoire construction is a preliminary step of any immunological analysis based on Rep-seq reads.
            Accurate construction of adaptive immune repertoire (antibody / TCR) avoids erroneous analysis of natural variations and antibody abundances.
        </p>

        <p align = justify>
            Repertoire construction problem can be formulated as a huge instance of error-correction or read clustering problem:
        </p>
        <table align = center width = "100%">
            <tr>
                <td align = center width = "10000000px">
                    <img src="docs/problem_antibody_repertoire_construction.svg" width = "80%" alt="pipeline">
                </td>
            </tr>
        </table>

        <h3>Input / output</h3>
        <p align = justify>
            A repertoire construction tool takes full-length Rep-seq reads as an input
            and constructs a repertoire as a set of <i>clusters</i>.
            Each cluster represent a group of reads corresponding to identical antibody (or TCRs) chains.
            Clusters in a repertoire are characterized by sequence and abundance.
            IgReC reports repertoires in
            <a href = "ig_repertoire_constructor_manual.html#repertoire_files">CLUSTERS.FASTA and RCM</a> formats.
        </p>

        <hr>

        <!-------------------------------------------------------------- --->

        <h2 id = "igrec_description">IgReC</h2>
        <p align = justify>
            IgReC takes as an input reads (paired-end or single) covering variable region of antibodies and
            corrects sequencing and amplification errors.
            Algorithm performs the following steps:
            <ul>
                <li><b>VJ Finder:</b> VJ labeling, cleaning and cropping input reads.</li>
                <li><b>Constructing & clustering Hamming graph:</b> representation of cleaned reads as a Hamming graph and finding of groups of highly similar antibodies.</li>
                <li><b>Receptor sequence construction:</b> reconstruction of antibody sequences from the constructed groups, cleaning amplification errors.</li>
            </ul>
        </p>

        <h3 id = "vj_finder">VJ Finder</h3>
        <p align = justify>
        The main goals of VJ Finder tool are:

        <ol>
            <li>Finding and discarding contaminated reads.</li>
            <li>Cropping remaining immunosequencing reads by the start position of the closest V gene segment and the end position of the closest J gene segment.
            VJ Finder also discards reads covering V gene segment only since they can not be unambiguously assigned to a cluster in a repertoire.</li>
        </ol>

        <p align = justify>
            VJ Finder outputs information about the closest V and J gene segments in tab-separated view:
        </p>

        <small>
        <table>
            <tr align="center">
                <td><b>Read id</b></td>
                <td><b>V start</b></td>
                <td><b>V end</b></td>
                <td><b>V score </br>(% identity)</b></td>
                <td><b>V id</b></td>
                <td><b>J start</b></td>
                <td><b>J end</b></td>
                <td><b>J score </br>(% identity)</b></td>
                <td><b>J id</b></td>
            </tr>
            <tr align="center">
                <td>read1</td> <td>1</td> <td>296</td> <td>100.0</td> <td>IGHV3-20*01</td> <td>321</td>
                <td>366</td> <td>89.0</td> <td>IGHJ5*02</td>
            </tr>
            <tr align="center">
                <td>read2</td> <td>1</td> <td>294</td> <td>98.64</td> <td>IGHV3-9*01</td> <td>309</td>
                <td>354</td> <td>100.0</td> <td>IGHJ2*01</td>
            </tr>
            <tr align="center">
                <td>...</td> <td>...</td> <td>...</td> <td>...</td> <td>...</td> <td>...</td>
                <td>...</td> <td>...</td> <td>...</td>
            </tr>
        </table>
        </small>


        <h3 id = "graph_algos">Constructing & clustering Hamming graph</h3>
        <p align = justify>
            IgReC constructs Hamming graph to identify similar Rep-seq reads.
            Vertices of Hamming graph represent unique reads.
            An edge connects <i>read1</i> and <i>read2</i> if Hamming distance between these them is relatively small.
        </p>
        <p align = justify>
            After construction of graph, IgReC launches <a href = "dense_subgraph_finder.html">DSF</a> for finding dense subgraphs.
            Dense subgraphs correspond to groups of highly similar antibodies:
        </p>

        <table align = center border = "0">
        <tr>
            <td align = center><img src="figures/graph_357134_default.jpg" alt = "hamming_graph_before_clusterization" width = 300px ></td>
            <td align = center><img src="figures/graph_357134_colored.jpg" alt = "hamming_graph_after_clusterization" width = 300px></td>
        </tr>
        <tr>
            <td>
                <small>Example of connected component of Hamming graph constructed from immunosequencing reads (heavy chain human repertoire).
                </small>
            </td>
            <td>
                <small>
                    Clustering of Hamming graph performed by <a href = "dense_subgraph_finder.html">DSF</a>.
                    Each color corresponds to antibody cluster in repertoire.
                </small>
            </td>
        </tr>
        </table>

        <!---
        <h3>Receptor sequence construction</h3>
        --->

        <hr>

        <!-------------------------------------------------------------- --->

    <h2 id = "ms_analyzer_description">Mass Spectra Analyzer</h2>
    <p align = "justify">
        This step performs immunoproteogenomics analysis to validate constructed repertoire using mass spectra.
        It takes alignment of mass spectra against constructed repertoire in
        <a href = "http://www.psidev.info/mzidentml">mzIdentML 1.1 format</a>
        (e.g., generated by <a href = "http://www.digitalproteomics.com/msgfplus.html">MS-GF+</a>) as an input and
        outputs a set of metrics and plots that show
        similarity between constructed repertoire and mass spectra:
    </p>

    <table align="center" border="0">
        <tr>
            <td><img width="100%" src="figures/ms_analysis/peptide_length_Total.png" alt="peptide_length_distribution"></td>
            <td><img width="100%" src="figures/ms_analysis/PSM_cov_Total.png" alt="psm_coverage_histogram"></td>
        </tr>
        <tr>
            <td align = center><small>Histogram of peptide length distribution.</small></td>
            <td>
                <small>Histogram of variable region coverage by PSMs.
                    Grey bars correspond to expected positions of CDRs.
                </small>
            </td>
        </tr>
    </table>
        <hr>

        <!-------------------------------------------------------------- --->

        <h2 id = "references">Manual and citations</h2>
        <p align = "center">
            <ul>
                <li>IgRepertoireConstructor manual can be found <a href = 'ig_repertoire_constructor_manual.html'>here</a>.</li>

            </ul>
        </p>

        <h3>Citations</h3>
        <p align = "justify">
        If you use our tools in your research, please cite our papers:
            <ul>
                <li>
                    <b>IgReC / BarcodedIgReC / IgQUAST / DiversityAnalyzer:</b>
                    Alexander Shlemov, Sergey Bankevich,
                    Andrey Bzikadze, Dmitry Chudakov, Yana Safonova, and Pavel A. Pevzner.
                    Reconstructing antibody repertoires from error-prone immunosequencing datasets.
                    Submitted.
                </li><br>
                <li>
                    <b>Immunoproteogenomics search / DSF:</b>
                    Yana Safonova, Stefano Bonissone,
                    Eugene Kurpilyansky, Ekaterina Starostina, Alla Lapidus,
                    Jeremy Stinson, Laura DePalatis, Wendy Sandoval, Jennie Lill, and Pavel A. Pevzner.
                    IgRepertoireConstructor: a novel algorithm for antibody repertoire construction and immunoproteogenomics analysis.
                    Bioinformatics. 2015 Jun 15; 31(12): i53-i61.
                    <a href = http://bioinformatics.oxfordjournals.org/content/31/12/i53.long>doi: 10.1093/bioinformatics/btv238.</a>
                </li>
            </ul>
        </p>

      <footer class="site-footer">
            <p align = justify>
                <big>You could also be interested in other immunogenomics tools created in our lab:</br>
                    <ul>
                        <li>
                            <a href = "barcodedIgReC.html">
                                <b>Barcoded IgReC, a modification of IgReC algorithm for barcoded Rep-seq data</b>
                            </a>
                        </li>
                        <li>
                            <a href = "diversity_analyzer.html">
                                <b>Diversity Analyzer, a tool for diversity analysis of full-length Rep-seq sequences</b>
                            </a>
                        </li>
                        <li>
                            <a href = "igquast.html">
                                <b>IgQUAST, a quality assessment tool for full-length adaptive immune repertoires</b>
                            </a>
                        </li>
                        <li>
                            <a href = 'http://yana-safonova.github.io/ig_simulator/'>
                                <b>IgSimulator, a versatile immunosequencing simulator</b>
                            </a>.
                        </li>
                    </ul>
                </big>
            </p>            
      </footer>

        <!---

      <footer class="site-footer">
        <span class="site-footer-owner">
            <a href="https://github.com/yana-safonova/ig_repertoire_constructor">IgRepertoireConstructor</a> is maintained by
            Center for Algorithmic Biotechnology, Saint Petersburg State University.
        </span>
      </footer>
      --->

    </section>

  
  </body>
</html>

