<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>IgRepertoireConstructor by yana-safonova</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">IgRepertoireConstructor</h1>
      <h2 class="project-tagline">An algorithm for error-correction of immunosequencing Illumina MiSeq reads and immunoproteogenomics analysis</h2>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor" class="btn">View on GitHub</a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/yana-safonova/ig_repertoire_constructor/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
    
    <p><b>IgRepertoireConstructor</b> performs error-correction of immunosequencing reads and uses mass spectra to validate the constructed antibody repertoires. IgRepertoireConstructor pipeline is presented below:</br>
    <table align = center>
    <tr><td align = center><img src="figures/pipeline.png" width = "80%" alt="pipeline"></td></tr>
    </table>
    </p>

    <p>Pipeline includes three main steps:
    <ul>
        <li><b>Illumina library preparation</b>. IgRepertoireConstructor takes Illumina MiSeq paired end reads as an input and expect that each left and right reads overlap and cover variable region of antibody. IgRepertoireConstructor stitches paired-end reads and removes contaminations using alignment of stitched reads against Ig germline database.</li>

        <li><b>Construction of antibody repertoire.</b> IgRepertoireConstructor takes as an input reads covering variable region of antibodies and corrects Illumina sequencing errors. Algorithm constructs Hamming graph on input reads and uses accurate clustering techniques for reconstruction of original antibody clusters:

        <table align = center border = "0">
        <tr> <td align = center><img src="figures/graph_357134_default.jpg" alt = "hamming_graph_before_clusterization" width = 400px ></td> <td align = center><img src="figures/graph_357134_colored.jpg" alt = "hamming_graph_after_clusterization" width = 400px></td> </tr>
        <tr><td>Example of connected component of Hamming graph constructed from immunosequencing reads (human, heavy chain repertoire of IgM).</td> <td>Clusterization of Hamming graph performed by IgRepertoireConstructor. Each color corresponds to single antibody.</td></tr>
        </table>
        </li>

        <li><b>Immunoproteogenomics analysis.</b> Constructed repertoire can be used as database for peptide identification from mass spectra. IgRepertoireConstructor takes as an input MZID file and computes various statistics about constructed repertoire.</li>
    </ul>
    </p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/yana-safonova/ig_simulator">IgRepertoireConstructor</a> is maintained by <a href="https://github.com/yana-safonova">Yana Safonova</a> (Center for Algorithmic Biology, Saint Petersburg State University).</span>

<!--        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span> 
-->
      </footer>

    </section>

  
  </body>
</html>

