<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
    <title>IgRepertoireConstructor 2.0 Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
    <style>
    </style>
</head>

<body>
<h1>IgRepertoireConstructor 2.0 manual</h1>
1. <a href="#intro">What is IgRepertoireConstructor?</a><br>

2. <a href="#install">Installation</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.1. <a href="#test_datasets">Verifying your installation</a><br>

3. <a href="#ig_repertoire_constructor_usage">IgRepertoireConstructor usage</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#repertoire_constructor_basic">Basic options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#repertoire_constructor_advanced">Advanced options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#repertoire_constructor_examples">Examples</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.4. <a href="#repertoire_constructor_output">Output files</a><br>

4. <a href = "#immunoproteogenomics_usage">MassSpectraAnalyzer usage</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;4.1. <a href = "#msanalyzer_basic">Basic options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;4.2. <a href = "#msanalyzer_advanced">Advanced options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;4.3. <a href = "#msanalyzer_examples">Examples</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;4.4. <a href = "#msanalyzer_output">Output files</a><br>

5. <a href = "#examples">Examples</a><br>

6.  <a href="#repertoire_files">Antibody repertoire representation</a><a><br>
&nbsp;&nbsp;&nbsp;&nbsp;6.1. <a href="#clusters_fasta">CLUSTERS.FASTA file format</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;6.2. <a href="#read_cluster_map">RCM file format</a><br>

7. <a href="#feedback">Feedback and bug reports</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;7.1. <a href="#citation">Citation</a><br>

<!--- ----------------- -->

<a id="intro"></a>
<h2>1. What is IgRepertoireConstructor?</h2>
IgRepertoireConstructor is a tool for construction of antibody repertoire and immunoproteogenomics analysis.
IgRepertoireConstructor includes two parts:
<ul>
    <li><b>IgRepertoireConstructor</b> &mdash; a tool for construction of antibody repertoire from cleaned Illumina Ig-Seq library.</li>
    <li><b>MassSpectraAnalyzer</b> &mdash; a tool for analysis of matching mass spectra against constructed repertoire.</li>
</ul>
IgRepertoireConstructor pipeline is shown below:<br>
<p align = center>
<img src="docs/ig_repertoire_constructor_pipeline.png" alt="pipeline" width = 500>
</p>

IgRepertoireConstructor pipeline by stages is shown below:<br>
<p align = center>
<img src="docs/ig_repertoire_constructor_pipeline_stages.png" alt="pipeline" width = 618>
</p>
<!--- ---------------------------------------------------------------- -->
<!--- ---------------------------------------------------------------- -->

<a id="install"></a>
<h2>2. Installation</h2>

IgRepertoireConstructor requires the following pre-installed dependencies:
<ul>
  <li>64-bit Linux system</li>
    <li>g++ (version 4.7 or higher)</li>
    <li>cmake (version 2.8.8 or higher)</li>
  <li>Python 2 (version 2.7 or higher), including:</li>
    <ul>
        <li><a href = "http://biopython.org/wiki/Download">BioPython</a></li>
        <li><a href = "http://matplotlib.org/users/installing.html">Matplotlib</a></li>
        <li><a href = "http://www.numpy.org/">NumPy</a></li>
        <li><a href = "http://www.scipy.org/install.html">SciPy<a></li>
        <!-- <li><a href = "http://www.scipy.org/install.html">Something else???? Alex, check it out<a></li> -->
    </ul>
  <li>zlib</li>
  <li>libbz2</li>
</ul>

To install IgRepertoireConstructor, type:
<pre class="code">
    <code>
    ./prepare_cfg; make
    </code>
</pre>

<a id="test_datasets"></a>
<h3>2.1. Verifying your installation</h3>
For testing purposes, IgRepertoireConstructor comes with several toy data sets. <br><br>

&#9658; To try IgRepertoireConstructor on the test data set, run:
<pre class="code"><code>
    ./ig_repertoire_constructor.py --test
</code>
</pre>

If the installation is successful, you will find the following information at the end of the log:

<pre class="code">
    <code>
    Thank you for using IgRepertoireConstructor!
    Log was written to igrepcon_test/ig_repertoire_constructor.log
    </code>
</pre>

Other test data set files can be found in the <code>test_dataset</code> directory.<br><br>

&#9658; To try MassSpectraAnalyzer on test data set, run:
<pre class="code">
<code>
    ./mass_spectra_analyzer.py --test
</code>
</pre>

If the installation is successful, you will find the following information at the end of the log:

<pre class="code">
    <code>
    Spectras processed: example_HC_chymo_CID.mzid.spectra, example_HC_trypsin_CID.mzid.spectra
    Metrics written to &lt;your_installation_directory>/ms_analyzer_test/metrics.txt
    Covered CRDs written to &lt;your_installation_directory>/ms_analyzer_test/covered_cdrs.txt
    PSM on IG regions written to &lt;your_installation_directory>/ms_analyzer_test/psm_on_ig_regions.txt
    Figures and statistics saved in &lt;your_installation_directory>/ms_analyzer_test
    </code>
</pre><br>

<!--- ---------------------------------------------------------------- -->
<!--- ---------------------------------------------------------------- -->

<a id="ig_repertoire_constructor_usage"></a>
<h2>3. IgRepertoireConstructor usage</h2>

To run IgRepertoireConstructor, type:
<pre class="code"><code>
    ./ig_repertoire_constructor.py [options] -s &lt;filename&gt; -o &lt;output_dir&gt;
</code>
</pre>

<!--- --------------------- -->

<a id="repertoire_constructor_basic"></a>
<h3>3.1. Basic options:</h3>

<code>-s &lt;filename&gt;</code><br>
FASTQ file with Illumina reads (required). <br><br>

<code>-o/--output &lt;output_dir&gt;</code><br>
Output directory (required). <br><br>

<code>-t/--threads &lt;int&gt;</code><br>
the number of parallel threads. The default value is 16.<br><br>

<code>--test</code><br>
Running on the toy test dataset. Command line corresponding to the test run is equivalent to the following:
<pre class = "code">
    <code>
    ./ig_repertoire_constructor.py -s test_dataset/merged_reads.fastq -o igrepcon_test
    </code>
</pre>

<code>--help</code><br>
printing help.<br><br>

<!--- --------------------- -->

<a id="repertoire_constructor_advanced"></a>
<h3>3.2. Advanced options:</h3>

<code>--tau &lt;int&gt;</code><br>
Maximum allowed number of mismatches between two reads corresponding to the same antibody. The default value is 4.<br>
<b>NOTE:</b> Be careful, using large value can significantly slow up processing or even can be inappropriate.
Reasonable value lies between 1 and 6.
<br><br>

<code>--chain/C &lt;str&gt;</code><br>
Ig chain type: <code>all</code> [default] (for both heavy and light chains) /
<code>heavy</code> / <code>light</code> (for both kappa and lambda chains) /
<code>lambda</code> / <code>kappa</code>
<br><br>

<code>--no-pseudogenes</code><br>
Do not use pseudogenes along with normal gene segments for VJ alignment
<br><br>

<code>--organism &lt;str&gt;</code><br>
Organism (<code>human</code> [defaut] and <code>mouse</code> only are supported for this moment)
<br><br>




<!--- --------------------- -->

<a id="repertoire_constructor_examples"></a>
<h3>3.3. Examples</h3>
To construct antibody repertoire from Illumina reads, covering variable regions of immunoglobulins
<pre class="code">
    <code>
    ./ig_repertoire_constructor.py -s reads.fastq -o output_dir
    </code>
</pre>
<!--- --------------------- -->

<a id="repertoire_constructor_output"></a>
<h3>3.4. Output files</h3>
IgRepertoireConstructor creates working directory (which name was specified using option <code>-o</code>) and outputs there the following files:

<ul>
  <li>Final repertoire files:</li>
  <ul>
    <li><b>final_repertoire.fa</b> &mdash; CLUSTERS.FASTA file with constructed repertoire (see details in <a href= "#repertoire_files">Antibody repertoire representation</a>).</li>
    <li><b>final_repertoire.rcm</b> &mdash; RCM file with constructed repertoire (see details in <a href = "#repertoire_files">Antibody repertoire representation</a>).</li>
    <li><b>final_repertoire_stripped.fa</b> &mdash; CLUSTERS.FASTA file with constructed repertoire filtered by component size (component size >= 5)</li>
    <li><b>supernodes.fa</b> &mdash; FASTA file with naive repertoire builded using supernodes (reads with abundance >= 5)</li>
  </ul>
  <li>VJ finder output files:</li>
  <ul>
    <li><b>vj_finder/cropped.fa</b> &mdash; FASTA file with initial reads, which were correctly reversed, aligned on V and J genes and cropped using VJ finder</li>
    <li><b>vj_finder/bad.fa</b> &mdash; FASTA file with initial reads, which were filtered by aligning on V and J genes</li>
    <li><b>vj_finder/add_info.csv</b> &mdash; CSV file with alignment info (consistent with <b>cropped.fa</b>); for each read it stores read ID, best hit V gene position, score and ID and the same for best hit J gene
  </ul>
  <li>Dense subgraph finder files:</li>
  <ul>
    <li><b>dense_subgraph_finder/dense_subgraph_finder.log</b> &mdash; log of DenseSubgraphFinder run</li>
    <li><b>dense_subgraph_finder/dense_subgraphs.txt</b> &mdash; dense subgraph decomposition in simple format: each line contains cluster label</li>
  </ul>
  <li>Auxiliary files:</li>
<ul>
    <li><b>ig_repertoire_constructor.log</b> &mdash; full log of IgRepertoireConstructor run</li>
    <li><b>compressed.fa</b> &mdash; FASTA file, which contains compressed reads, result of IgTrieCompressor; each cropped read collapsed to minimal presented prefix.
      Thus it's guarantied that each read is not prefix of any other read, only unique prefixes are extracted.
      For each unique prefix-read its <i>abundance</i> (the number of joined reads) is added to its ID
    </li>
    <li><b>map.txt</b> &mdash; file, which stores read compression map provided by igTrieCompressor.
      This file is consistent with <b>cropped.fa</b>, each line contains the number (zero based) of target prefix-read in <b>compressed.fa</b> file</li>
    <li><b>sw.graph</b> &mdash; Smith-Waterman graph constructed on compressed reads from <b>compressed.fa</b> in METIS *.graph format (see <a href="http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/manual.pdf">METIS 5.1.x Manual</a> for format description)
</ul>
</ul>
<br>

<!-- -------------------------------------------------------------------- -->
<!-- -------------------------------------------------------------------- -->

<a id="immunoproteogenomics_usage"></a>
<h2>4. MassSpectraAnalyzer usage</h2>
MassSpectraAnalyzer takes as an input result of matching of mass spectra against the constructed repertoire in <a href = "http://www.psidev.info/mzidentml">mzIdentML 1.1 format</a> (e.g., generated by <a href = "http://www.digitalproteomics.com/msgfplus.html">MS-GF+</a>) and computes multiple statistics showing coverage of the constructed repertoire by mass spectra. </br></br>

&#9658; To run MassSpectraAnalyzer, type:
<pre class="code">
<code>
    ./mass_spectra_analyzer.py [options] -o &lt;output_dir> input_file_1.mzid ... input_file_N.mzid
</code>
</pre>

<!-- --------------------- -->
<a id = "msanalyzer_basic"></a>
<h3>4.1. Basic options:</h3>

<code>inputs</code></br>
Input files with mass spectra alignment to protein database in <a href = "http://www.psidev.info/mzidentml">mzIdentML 1.1 format</a>.</br></br>

<code>-o &lt;output_dir></code></br>
output directory (required).</br></br>

<code>--test</code></br>
Running on the toy test data set. <!-- Command line corresponding to the test run is equivalent to the following:
<pre class = "code">
    <code>
    ./mass_spectra_analysis.py --regions test_dataset/MS_analysis/example_regions.txt test_dataset/MS_analysis/example_HC_chymo_CID.mzid.spectra test_dataset/MS_analysis/example_HC_trypsin_CID.mzid.spectra
    </code>
</pre> --></br></br>

<code>--help, -h</code></br>
Printing help.</br>

<!-- --------------------- -->
<a id = "msanalyzer_advanced"></a>
<h3>4.2. Advanced options:</h3>
<code>--regions &lt;filename></code></br>
File with information about framework and CDRs for protein sequences from used database in IgBLAST format.
Example of file with labeled regions is given below:
<pre class = "code">
    <code>
    Query= Antibody_sequence_1
    CDR2-IMGT       51      58
    FR2-IMGT        34      50
    FR1-IMGT        1       25
    FR3-IMGT        59      94
    CDR1-IMGT       26      33
    CDR3-IMGT       95      110
    Query= Antibody_sequence_2
    CDR2-IMGT       51      58
    FR2-IMGT        34      50
    FR1-IMGT        1       25
    FR3-IMGT        59      94
    CDR1-IMGT       26      33
    CDR3-IMGT       95      111
    </code>
</pre>
where Antibody_sequence_1 and Antibody_sequence_2 are sequences from database.

<!-- --------------------- -->
<a id = "msanalyzer_examples"></a>
<h3>4.3. Examples:</h3>

&#9658; To compute statistics for both chymo and trypsin mass spectra datasets and labeled regions, run the following command:
<pre class = "code">
    <code>
    ./mass_spectra_analysis.py --output output_dir --regions regions.align example_HC_chymo_CID.mzid example_HC_trypsin_CID.mzid
    </code>
</pre>

<!-- --------------------- -->
<a id = "msanalyzer_output"></a>
<h3>4.4. Output files:</h3>
<ul>
    <li>Statistics:</li>
    <ul>
        <li><b>metrics.txt</b> - file with basic statistics for each of given mass spectrum alignments.</li>
        <li><b>covered_cdrs.txt</b> - file with information about number of sequences with at least one peptide spectrum match on corresponding region.</li>
        <li><b>psm_on_ig_regions.txt</b> - file with information about number of peptide spectrum matches aligned to corresponding regions of sequences.</li>
    </ul></br>

    <li>Statistics visualization: </li>
    <ul>
        <li><b>PSM_cov.png</b> - PNG file with plot showing coverage by peptide spectrum matches along antibody sequence.</li>
        <li><b>peptide_cov.png</b> - PNG file with plot showing coverage by peptide spectrum matches along antibody sequence, consider only scans with unique alignment to database.</li>
        <li><b>PSM_per_scan.png</b> - PNG file with histogram of distribution of number of peptide spectrum matches per scan.</li>
        <li><b>peptide_length.png</b> - PNG file with histogram of distribution of peptide length.</li>
    </ul>
</ul>

<!-- -------------------------------------------------------------------- -->

<a id="examples"></a>
<h2>5. Examples</h2>
Example shows IgRepertoireConstructor pipeline in action for merged paired-end Illumina MiSeq library including reads <b>reads.fastq</b> and mass spectra <b>AspN_CID.mzXML</b> corresponding to the same antibody repertoire.<br><br>
&#9658; To run IgRepertoireConstructor with standard settings, type the following command:
<pre class = "code">
    <code>
    ./ig_repertoire_constructor.py -s reads.fastq -o repertoire_constructing
    </code>
</pre>
Sequences of the constructed repertoire are located in <b>repertoire_constructing/constructed_repertoire.clusters.fa</b>. They can be converted into amino acid sequences and used as a database for matching mass spectra <b>AspN_CID.mzXML</b> (e.g., using MS-GF+ tool). Let result of MS-GF+ tool be a file <b>AspN_CID.mzId</b>.<br><br>

&#9658; To run MassSpectraAnalyzer on <b>AspN_CID.mzId</b> file, type the following command:
<pre class = "code">
    <code>
    ./mass_spectra_analyzer.py -o ms_analysis AspN_CID.mzId
    </code>
</pre>
Statistics for mass spectra alignment can be found in <b>ms_analysis</b> directory.
<br><br>

<!-- -------------------------------------------------------------------- -->

<a id="repertoire_files"></a>
<h2>6. Antibody repertoire representation</h2>
We used two files for representation of repertoire for the set of clustered reads: CLUSTERS.FASTA and RCM.

<a id="clusters_fasta"></a>
<h3>6.1. CLUSTERS.FASTA file format</h3>
CLUSTERS.FASTA is a FASTA file, where sequences correspond to the assembled antibodies.
Each header contains information about corresponding antibody cluster (id and size):
<pre class="code">    <code>
    &gt;cluster___1___size___3
    CCCCTGCAATTAAAATTGTTGACCACCTACATACCAAAGACGAGCGCCTTTACGCTTGCCTTTAGTACCTCGCAACGGCTGCGGACG
    &gt;cluster___2___size___2
    CCCCTGCAATTAAAATTGTTGACCACCTACATACCAAAGACGAGCGCCTTTACGCTTGCCTTTAGTACCTCGCAACGGCTGCGG
    &gt;cluster___3___size___1
    CCCCTGCAATTAAAATTGTTGACCACCTACATACCAAAGACGAGCGCCTTTACGCTTGCCTTTAGTACCTCGCAACGGCTGCGGAC
    </code>
</pre>

<a id="read_cluster_map"></a>
<h3>6.2. RCM file format</h3>
Every line of RCM (read-cluster map) file contains information about read name and corresponding cluster ID:
<pre class="code">    <code>
    MISEQ@:53:000000000-A2BMW:1:2114:14345:28882    1
    MISEQ@:53:000000000-A2BMW:1:2114:14374:28884    1
    MISEQ@:53:000000000-A2BMW:1:2114:14393:28886    1
    MISEQ@:53:000000000-A2BMW:1:2114:16454:28882    2
    MISEQ@:53:000000000-A2BMW:1:2114:16426:28886    2
    MISEQ@:53:000000000-A2BMW:1:2114:15812:28886    3
    </code>
</pre>

<br>
Reperoire described in the example above consists of three antibodies. E.g., the antibody with ID 1 has abundancy 3, since it was constructed from three reads:<br>
MISEQ@:53:000000000-A2BMW:1:2114:14345:28882<br>
MISEQ@:53:000000000-A2BMW:1:2114:14374:28884<br>
MISEQ@:53:000000000-A2BMW:1:2114:14393:28886<br><br>
<b>NOTE:</b> IDs in CLUSTERS.FASTA and RCM files are consistent.
<br><br>

<!--- -------------------------------------------------------------------- --->
<a id="feedback"></a>
<h2>7. Feedback and bug reports</h2>
Your comments, bug reports, and suggestions are very welcome.
They will help us to further improving on IgRepertoireConstructor.
<br><br>
If you have any troubles with running IgRepertoireConstructor, please send us the log file from the output directory.
<br><br>
Address for communications: <a href="mailto:igtools_support@googlegroups.com">igtools_support@googlegroups.com</a>.

<a id = "citation"></a>
<h3>7.1. Citation</h3>
If you use IgRepertoireConstructor in your research, please refer to <a href="http://bioinformatics.oxfordjournals.org/content/31/12/i53.long" target="_blank">Safonova et al., 2015</a>.
</body></html>
