<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
    <title>Barcode QUAST Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
    <style>
    </style>
</head>

<body>
<h1>BarcodeQUAST manual</h1>
1. <a href="#intro">What is BarcodeQUAST?</a><br>

2. <a href="#install">Installation</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;2.1. <a href="#test_datasets">Verifying your installation</a><br>

3. <a href="#usage">Usage</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#basic_options">Basic options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#advanced_options">Advanced options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.3. <a href="#examples">Examples</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.4. <a href="#output">Output files</a><br>

4. <a href="#metrics">Metrics</a><br>

<!--- ----------------- --->

<a id="intro"></a>
<h2>1. What is BarcodeQUAST?</h2>
BarcodeQUAST is application for quality evaluation of a constructed repertoire in comparison with an ideal repertoire. <br>
As an ideal repertoire you can use, for example, assembly of a sample, prepared using unique molecular identifiers(barcodes) to trace individual cDNA molecules. Using information about barcodes, near error-free assembly can be done with <a href="http://milaboratory.com/projects/migec-project/">MiGEC</a> tool and it can be used as an ideal repertoire. Also you can use simulated data from <a href="http://yana-safonova.github.io/ig_simulator/">IgSimulator</a> tool.

<!--- ---------------------------------------------------------------- --->
<!--- ---------------------------------------------------------------- --->

<a id="install"></a>
<h2>2. Installation</h2>

BarcodeQUAST supplied as a part of IgRepertoireConstructor tool and require same dependencies plus some additional Python libs. Full list of dependencies listed below:
<ul>
	<li>64-bit Linux system</li>
    <li>g++ (version 4.7 or higher)</li>
    <li>cmake (version 2.8.8 or higher)</li>
	<li>Python (version 2.7 or higher), including:</li>
    <ul>
        <li><a href = "http://biopython.org/wiki/Download">BioPython</a></li>
        <li><a href = "http://matplotlib.org/users/installing.html">Matplotlib</a></li>
        <li><a href = "http://www.numpy.org/">NumPy</a></li>
        <li><a href = "http://www.scipy.org/install.html">SciPy<a></li>
    </ul>
	<li>zlib</li>
    <li>libbz2</li>
</ul>

To install BarcodeQUAST, type:
<pre class="code">
    <code>
    ./prepare_cfg
	make
    </code>
</pre>


<!--- ---------------------------------------------------------------- --->

<a id="usage"></a>
<h2>3. Usage</h2>

To run BarcodeQUAST, type:
<pre class="code"><code>
    ./barcode_quast.py [options] --Bc &lt;barcode_clusters&gt; -c &lt;data_clusters&gt; --out &lt;output_dir&gt;
</code>
</pre>

<!--- --------------------- --->

<a id="basic_options"></a>
<h3>3.1. Basic options:</h3>

<code>-s &lt;filename&gt;</code><br>
FASTQ file with Illumina reads (required). <br><br>

<code>-h/--help</code><br>
show help message<br><br>

<code>--Bc &lt;barcode_clusters&gt;</code><br>
Barcode sequences assembly in CLUSTERS.FA or MiGEC format (required).<br><br>

<code>--Br &lt;barcode_clusters&gt;</code><br>
RCM(read-cluster mapping) file for barcodes assembly.<br><br>

<code>-c &lt;data_clusters&gt;</code><br>
CLUSTERS.FA file for assembly to be evaluated (required).<br><br>

<code>-r &lt;data_rcm&gt;</code><br>
RCM file for assembly to be evaluated (required).<br><br>

<code>--threads-num &lt;threads_num&gt;</code><br>
Number of threads to use (by default 4).<br><br>

<code>--out&lt;output_dir&gt;</code><br>
Output directory (required).<br><br>

<!--- --------------------- --->

<a id="advanced_options"></a>
<h3>3.2. Advanced options:</h3>

<code>--tau &lt;tau&gt;</code><br>
Maximum allowed number of mismatches/gaps between sequences in barcodes and evaluated repertoire (by default 3).<br><br>

<code>--rate &lt;rate&gt;</code><br>
Rate to distinguish between good and bad components (by default 0.5).<br><br>

<code>--deep-rcm-cmp</code><br>
Align clusters from evaluated repertoire against corresponding barcode sequence and compute stats for such alignments. Computed only for good barcodes, recommended to with for only very high rate value (about 0.9) because of slow speed.<br><br>

<code>--rerun</code><br>
Run with match results from the output directory.<br><br>

<!--- --------------------- --->

<a id="examples"></a>
<h3>3.3. Examples</h3>
To evaluate constructed repertoire with barcodes
<pre class="code">
    <code>
    ./barcode_quast.py --Bc barcodes.fastq -c constructed_clusters.fa -o output_dir
    </code>
</pre>
<!--- --------------------- --->

<a id="output"></a>
<h3>3.4. Output files</h3>
BarcodeQUAST creates working directory (which name was specified using option <code>--out</code>) and outputs there the following files:
<ul>
    <li><b>metrics.txt</b> - file with computed metrics.</li>
    <li><b>cluster_sizes_distr.png</b> - graph with distribution of cluster sizes (number of reads with the same barcode/in one constructed cluster).</li>
    <li><b>cluster_sizes_distr_nt.png</b> - graph with distribution of cluster sizes (number of reads with the same barcode/in one constructed cluster) for non-trivial clusters (size &gt; 1).</li>
    <li><b>isolated_cluster_sizes_distr.png</b> - graph with distribution of cluster sizes (number of reads with the same barcode/in one constructed cluster) for isolated clusters (isolated cluster is a cluster that don't have a match in the second repertoire, a match is a cluster with sequence on distance &lt;= tau).</li>
    <li><b>isolated_cluster_sizes_distr_nt.png</b> - graph with distribution cluster sizes for isolated non-trivial clusters.</li>
	<li><b>lengths_distr.png</b> - graph with cluster sequence lengths distribution</li>
	<li><b>isolated_lengths_distr.png</b> - graph with sequence lengths distribution for isolated clusters</li>
    <li><b>barcode_quast.log</b> - full log of barcodeQUAST run.</li>
</ul>
<br>

<!--- ---------------------------------------------------------------- --->

<a id="metrics"></a>
<h2>2. Metrics</h2>
Metrics computed for each provided repertoire:
<ul>
    <li><b>#clusters</b> - number of clusters (sequences) in repertoire.</li>
    <li><b>#isolated clusters</b> - number of clusters in repertoire that don't have similar cluster in the second repertoire (isolated cluster is a cluster that don't have a match in the second repertoire, a match is a cluster with sequence on distance &lt;= tau).</li>
    <li><b>#clusters (&gt;= sz)</b> - number of clusters of size &gt;= sz (size of cluster is number of reads included to this cluster) in repertoire.</li>
    <li><b>Min sequence length</b> - minimal cluster sequence length.</li>
    <li><b>Avg sequence length</b> - average cluster sequence length.</li>
    <li><b>Max sequence length</b> - maximal cluster sequence length.</li>
    <li><b>Min size of cluster</b> - minimal cluster size.</li>
    <li><b>Avg size of cluster</b> - average cluster size.</li>
    <li><b>Max size of cluster</b> - maximal cluster size.</li>
    <li><b>Min size of cluster</b> - minimal size of isolated cluster.</li>
    <li><b>Avg size of cluster</b> - average size of isolated cluster.</li>
    <li><b>Max size of cluster</b> - maximal size of isolated cluster.</li>
</ul>

General metrics:
<ul>
    <li><b>#good barcodes</b> - number of barcodes having &gt;= rate * barcode_size reads included to the same cluster in evaluated repertoire (barcodes covered by a single cluster in evaluated repertoire).
      This metric shows quality of reads clusterization in evaluated repertoire.</li>
    <li><b>#bad barcodes</b> - (#all barcodes - #good barcodes)</li>
    <li><b>Min gap distance to good barcodes</b> - minimal gap distance to good barcode computed among all clusters that share reads with this barcode.
      Cluster that covers &gt; rate portion of reads in good barcode is excluded from the comparison.
      This metric is computed only if flag --deep-rcm-cmp is on.</li>
    <li><b>Avg gap distance to good barcodes</b> - average gap distance to good barcode computed among all clusters that share reads with this barcode.
      Cluster that covers &gt; rate portion of reads in good barcode is excluded from the comparison.
      This metric is computed only if flag --deep-rcm-cmp is on.</li>
    <li><b>Max gap distance to good barcodes</b> - maximal gap distance to good barcode computed among all clusters that share reads with this barcode.
      Cluster that covers &gt; rate portion of reads in good barcode is excluded from the comparison.
      This metric is computed only if flag --deep-rcm-cmp is on.</li>
    <li><b>Min mismatches distance to good barcodes</b> - minimal mismatches distance to good barcode computed among all clusters that share reads with this barcode.
      Cluster that covers &gt; rate portion of reads in good barcode is excluded from the comparison.
      This metric is computed only if flag --deep-rcm-cmp is on.</li>
    <li><b>Avg mismatches distance to good barcodes</b> - average mismatches distance to good barcode computed among all clusters that share reads with this barcode.
      Cluster that covers &gt; rate portion of reads in good barcode is excluded from the comparison.
      This metric is computed only if flag --deep-rcm-cmp is on.</li>
    <li><b>Max mismatches distance to good barcodes</b> - maximal mismatches distance to good barcode computed among all clusters that share reads with this barcode.
      Cluster that covers &gt; rate portion of reads in good barcode is excluded from the comparison.
      This metric is computed only if flag --deep-rcm-cmp is on.</li>

</ul>

</body></html>
